test/puppetlabs/jdbc_util/pool_test.clj

```diff
@@ -167,3 +167,25 @@
           (pool/connection-pool-with-delayed-init datasource {:migration-db migration-db-spec} identity 5000)
           (let [migrate-with (deref !migrate-with 5000 nil)]
             (is (= migration-db-spec migrate-with))))))))
+
+(deftest migration-blocks-initilization
+  (let [config (-> core-test/test-db
+                   (assoc :pool-name "AppPool")
+                   pool/spec->hikari-options
+                   pool/options->hikari-config)
+        num-migrations-left (fn []
+                              (count (migration/uncompleted-migrations core-test/test-db
+                                                                       "test-migrations")))]
+    (core/drop-public-tables! core-test/test-db)
+    (testing "waiting on migrations blocks startup on replicas"
+      (is (= 1 (num-migrations-left)))
+      (pool/connection-pool-with-delayed-init config
+                                              {:migration-db core-test/test-db
+                                               :migration-dir "test-migrations"
+                                               :replication-mode :replica}
+                                              (fn [_] nil)
+                                              5000)
+      (is (= 1 (num-migrations-left)))
```

Comment:
I'm not sure this tests what you expect.  If the goal is to test that the init blocked until after the migrations are done somewhere else, then it seems like we need to know that the init function was run when we expect and not run before we expect it to.  Maybe the local init function can impact a scoped atom to reflect the state change that can be tested against?